src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -296,29 +535,193 @@ private void PutCounterImpl(DocumentsOperationContext context, string documentId
+                    localValuePtr->Etag = sourceValue.Etag;
+
+                    value += sourceValue.Value;
+                    continue;
                 }
+
+                // counter doesn't have this dbId
+                modified = true;
+                value += sourceValue.Value;
+                var scope = AddPartialValueToExistingCounter(context, data, counterName, existingCounter, localDbIdIndex, sourceValue.Value, sourceValue.Etag);
+                _scopes.Add(scope);
             }
+
+            return value;
+        }
+
+        private static ByteStringContext<ByteStringMemoryCache>.InternalScope AddPartialValueToExistingCounter(DocumentsOperationContext context, BlittableJsonReaderObject data, string counterName,
+            BlittableJsonReaderObject.RawBlob existingCounter, int dbIdIndex, long sourceValue, long sourceEtag)
+        {
+            var scope = context.Allocator.Allocate((dbIdIndex + 1) * SizeOfCounterValues, out var newVal);
+
+            Memory.Copy(newVal.Ptr, existingCounter.Ptr, existingCounter.Length);
+            var empties = dbIdIndex - existingCounter.Length / SizeOfCounterValues - 1;
+            if (empties > 0)
+            {
+                Memory.Set(newVal.Ptr + existingCounter.Length, 0, empties * SizeOfCounterValues);
+            }
+;
+            var newEntry = (CounterValues*)newVal.Ptr + dbIdIndex;
+            newEntry->Value = sourceValue;
+            newEntry->Etag = sourceEtag;
+
+            data.Modifications = data.Modifications ?? new DynamicJsonValue(data);
+            data.Modifications[counterName] = new BlittableJsonReaderObject.RawBlob
+            {
+                Length = newVal.Length, Ptr = newVal.Ptr
+            };
+
+            return scope;
+
+        }
+
+        private static int[] SortValuesByDbIdIndex(Dictionary<LazyStringValue, int> dict,
+            BlittableJsonReaderArray localDbIds, BlittableJsonReaderArray sourceDbIds, int sourceCount)
+        {
+            var sorted = new int[dict.Count + sourceDbIds.Length];
+            Array.Fill(sorted, -1);
+
+            for (int i = 0; i < sourceCount; i++)
+            {
+                var dbId = (LazyStringValue)sourceDbIds[i];
+
+                if (dict.TryGetValue(dbId, out var dbIdIndex))
+                {
+                    sorted[dbIdIndex] = i;
+                }
+                else
+                {
+
+                    localDbIds.Modifications = localDbIds.Modifications ?? new DynamicJsonArray();
+                    localDbIds.Modifications.Add(dbId);
+
+                    var newIndex = dict.Count;
+                    sorted[newIndex] = i;
+
+                    dict.Add(dbId, newIndex);
+                }
+            }
+
+            return sorted;
+        }
+
+        private static int FindMaxDbIdIndex(int sourceCount, Dictionary<LazyStringValue, int> dict, BlittableJsonReaderArray localDbIds, BlittableJsonReaderArray sourceDbIds)
```

Comment:
I guess that this function can be merged with SortValuesByDbIdIndex, given that it won't need to be sorting anything, but just intializing the dict and modifying the localDbIds

Reply:
removed this function